gh workflow run

vilmibm opened the pull request:
This PR implements `gh workflow run`, a command for creating `workflow_dispatch` events for workflows that support them.

It accepts workflow input in one of four ways:

- JSON via STDIN
- `-f` and `-F` flags that mimic `gh api` raw/magic fields
- dynamic input prompts (video demo: https://youtu.be/oTAoNvbT9Tk )

If no arguments are provided, the user can interactively select an enabled workflow and 
then interactively provide inputs.

If any arguments are present, no interactivity is assumed.

The workflow file may be run from a different ref than the repo's default branch via `--ref`.

samcoe commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+	return out, nil
+}
+
+func getWorkflowContent(client *api.Client, repo ghrepo.Interface, workflow *shared.Workflow, ref string) ([]byte, error) {
```
Now that workflow view is merged, which also uses this exact function, we should share it between these commands.

samcoe commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+							inputsMapNode = node
+							break
+						}
+						if node.Value == "inputs" {
+							inputsKeyNode = node
+						}
+					}
+					break
+				}
+				if node.Value == "workflow_dispatch" {
+					dispatchKeyNode = node
+				}
+			}
+			break
+		}
+		if strings.EqualFold(node.Value, "on") {
+			onKeyNode = node
+		}
+	}
+
+	if onKeyNode == nil {
+		return nil, errors.New("invalid workflow: no 'on' key")
+	}
+
+	if dispatchKeyNode == nil {
+		return nil, errors.New("unable to manually run a workflow without a workflow_dispatch event")
+	}
+
+	out := map[string]WorkflowInput{}
+
+	if inputsKeyNode == nil || inputsMapNode == nil {
+		return out, nil
+	}
+
+	err = inputsMapNode.Decode(&out)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode workflow inputs: %w", err)
+	}
+
+	return out, nil
```
What about something like
```go
func findNode(root *yaml.Node, value string) (*yaml.Node, error) {
	for _, node := range root.Content {
		if node.Content == value {
			return node, nil
		}
	}
	return nil, errors.New("node not found")
}

onKeyNode, err := findNode(rootNode.Content[0].Content, "on")
if err != nil {
	return nil, errors.New("invalid workflow: no 'on' key")
}

dispatchKeyNode, err := findNode(onKeyNode, "workflow_dispatch")
if err != nil {
	return nil, errors.New("unable to manually run a workflow without a workflow_dispatch event")
}

out := map[string]WorkflowInput{}
inputKeyNode, err := findNode(dispatchKeyNode, "inputs")
if err != nil {
	return out, nil
}

err = inputKeyNode.Content.Decode(&out)
if err != nil {
	return nil, fmt.Errorf("could not decode workflow inputs: %w", err)
}

return out, nil
```

vilmibm replied:
This is a nice suggestion but doesn't cover the yaml syntax issue that mislav mentioned; I'll just keep the gnarly thing I have now

samcoe commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+			}
+		}
+
+		if len(opts.InputArgs) > 0 {
+			fs := pflag.FlagSet{}
+			for inputName, input := range inputs {
+				fs.String(inputName, input.Default, input.Description)
+			}
+			err = fs.Parse(opts.InputArgs)
+			if err != nil {
+				return fmt.Errorf("could not parse input args: %w", err)
+			}
+			for inputName := range inputs {
+				providedValue, _ := fs.GetString(inputName)
+				providedInputs[inputName] = providedValue
+			}
+		}
```
This turned out cleaner than I expected 👍

samcoe commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+	providedInputs map[string]string
+}
+
+func (ia *InputAnswer) WriteAnswer(name string, value interface{}) error {
```
I might be missing something but I don't see this being used anywhere, is it possible that it was meant to be deleted?

vilmibm replied:
It's called by `survey`

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+		return fmt.Errorf("unable to fetch workflow file content: %w", err)
+	}
+
+	inputs, err := findInputs(yamlContent)
```
It worries me that we have to fetch the workflow contents and parse the inputs for the `workflow_dispatch` event even in the case where a script supplied all necessary inputs to `gh workflow run` via command-line. If everything was already supplied by the user and we are not in interactive mode, why perform any client-side validation of the inputs? Instead we could send the payload to the server and let it error out on invalid inputs.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+			$ gh workflow run triage.yml --ref=myBranch
+
+			# Run the workflow file 'triage.yml' with command line inputs
+			$ gh workflow run triage.yml -- --name=scully --greeting=hello
```
Since `--name` and `--greeting` aren't real flags to `workflow run` nor are they forwarded to another command, would we consider an alternate syntax for supplying arbitrary key-value pairs? We already have the `-f/-F` syntax for supplying arbitrary parameters in `gh api`. The benefit of lifting that syntax would be that we could also support reading contents from file, e.g. `-F body=@mybody.md`, and that we wouldn't have to dynamically construct a temporary FlagSet to parse into. Another benefit is that for input names that have underscores in them, we avoid flag names that look like `--snake_case=`. Finally, we wouldn't have to require using `--` as separator.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+	// TODO this is pretty hideous
+	for _, node := range rootNode.Content[0].Content {
+		if onKeyNode != nil {
+			for _, node := range node.Content {
```
This looks like it might fail to parse `on: workflow_file`, which I believe is valid workflow syntax

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+
+	states := []shared.WorkflowState{shared.Active}
+	workflow, err := shared.ResolveWorkflow(
+		opts.IO, client, repo, opts.Prompt, opts.Selector, states)
```
It feels slightly wrong to me that the user who hasn't specified the workflow ID will be presented with the list of workflows to choose from, but the listed workflows aren't guaranteed to be `workflow_dispatch`-enabled. But, I find this acceptable since right now it would take N+1 queries to verify for each workflow if they have a workflow_dispatch trigger.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+
+	inputs, err := findInputs(yamlContent)
+	if err != nil {
+		return err
```
If we failed parsing inputs, couldn't we just print the warning on stderr but allow the command to continue? It might be that the workflow_dispatch trigger doesn't take any inputs. The rationale is that Actions workflow syntax might get updated in the future, and if we're too tied to it in the client, dispatching runs will be blocked on the client due to choking on unrecognized syntax.

mislav reviewed (changes requested):
I appreciate the amount of work you've put into this, I'm fearing that this might be over-engineered. My main gripes are about the brittleness of YAML parsing—especially because it affects non-interactive mode—and the syntax for specifying inputs.

In my testing, depending on the contents of the workflow file, I was able to dispatch a simple workflow using direct API access:
```
$ gh api repos/mislav/playground/actions/workflows/main.yml/dispatches -f ref=main
```
but not via this command:
```sh
# YAML parser was unable to detect the workflow_dispatch trigger
$ ghd workflow run -R mislav/playground main.yml
unable to manually run a workflow without a workflow_dispatch event

# "logLevel" is required but it also has a default value
$ ghd workflow run -R mislav/playground main.yml
missing required input 'logLevel'
```

I would even be down for _not_ offering the user to fill in inputs interactively until we can get the information about accepted workflow inputs from the API. Parsing YAML workflow files via fetching repository contents over API feels intrusive into the implementation detail of Actions.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+	}
+
+	for inputName, inputValue := range inputs {
+		if inputValue.Required && providedInputs[inputName] == "" {
```
The input might be marked as "required", but if it has a default value then it's not required to be filled in by the user.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,382 @@
+			$ echo '{"name":"scully", "greeting":"hello"}' | gh workflow run triage.yml
+
+			# Run the workflow file 'triage.yml' with JSON via --json
+			$ gh workflow run triage.yml --json='{"name":"scully", "greeting":"hello"}'
```
I like the ability to pass JSON via stdin! But then I don't think that we need an explicit JSON flag. Since JSON is machine data and likely will be constructed by tools rather than by hand, we should assume that it can be large. Passing potentially large datasets via argument should generally be discouraged since it can trigger the OS maximum limit for process arguments.

vilmibm commented:
@mislav I hear your concerns about over engineering / brittleness. I knew I was putting a lot up here and didn't expect it to necessarily all be merged.

I'm going to sleep on what exactly I think should be taken out, but here are some unordered thoughts:

- I've already patched the two YAML parsing issues and am not sure how to evaluate the risk of remaining brittleness. I wasn't aware of the `on: workflow_dispatch` form of the syntax, but I also am not sure there there are other forms beyond that and the `MappingNode` form which I'm now handling (though I haven't pushed yet because I was going to add relevant tests)
- I'm with you on not supporting `--json`; it makes sense to me. I didn't have a strong usecase and did both that and STDIN because I wasn't sure which was better.
- I'm not thrilled about `--` and agree that the presence of names like `underscore_arg` is a smell that it's not the right direction. I generally avoid the `-F field=value` concept because I find _that_ confusing and brittle, especially when escaping is concerned. That said, I think that's a personal problem and given that we already have -F in `api` and this feels like a similar usecase to that I'm fine with it. I also can forsee people wanting to use files as input which is unwieldy in the current state of the PR.
- I am not overly concerned with tying ourselves to the current actions syntax. Even should it change, it would be entirely possible to still use this command if I just forgo input validation for the JSON on STDIN case. I also feel like running workflows in general assumes a knowledge of workflow file syntax, but that might be an incorrect assumption.

So right now my TODO list for tomorrow is definitely:

- [ ] remove `--json`
- [ ] change from `--` input arguments to `-F` style

and in the morning I hope to come to a decision on:

- YAML parsing / up front input validation
- Whether or not the interactive input mode is actually something users want or if I just think it's neat and that's clouding my judgement

@samcoe @billygriffin @ampinsk any of y'all have thoughts on this?

vilmibm commented:
I've slept on it and talked to some folks and where I've landed is:

- if JSON is passed on STDIN _or_ `-F` is used, send it directly to the API without doing any YAML parsing
- In the no argument case (ie blank invocation or just a workflow filename), continue to parse YAML and collect interactive input, failing if the YAML can't be parsed

I recognize that the dynamic input collection is not strictly necessary to ship, but it's something I care about seeing in the release. My hope is that by making the JSON and `-F` input modes not contingent on YAML parsing then we reach a good middle ground of brittleness mitigation + `gh` style human-friendly featureness.

vilmibm committed:
review feedback

- Switch to -f/-F instead of -- (this lead to some `gh api` copypasta)
- Remove --json
- Don't parse workflow YAML if not collecting interactively
- Share GetWorkflowContent
- Fix a parsing issue

Commit:
share workflow content getting code

Commit:
gh workflow run

samcoe reviewed (approved):
This looks great! I like the thorough tests. There is lots of code copied from the API here that would be good to figure out a way to share in the future, especially the `-f/-F` flag processing which seems super useful.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+			$ gh workflow run triage.yml --ref=myBranch
+
+			# Run the workflow file 'triage.yml' with command line inputs
+			$ gh workflow run triage.yml -fname=scully -fgreeting=hello
```
Even though a space after `-f` isn't necessary, let's add it here for the sake of readability. People who are experience with shorthand flags will know that they can omit the space.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+			$ gh workflow run triage.yml
+
+			# Run the workflow file 'triage.yml' at a specified ref, interactively providing inputs
+			$ gh workflow run triage.yml --ref=myBranch
```
Nit: it's a bit weird to see camelCase branch names when that isn't very common in the world. How about `my-branch`?

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+	}
+
+	cmd := &cobra.Command{
+		Use:   "run [<workflow ID> | <workflow name>]",
```
Nit: we used to avoid spaces in placeholder names in favor of dashes. How would you feel about `<workflow-ID>`, `<workflow-name>`

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+
+			- Interactively
+			- via -f or -F flags
+			- As JSON, via STDIN
```
Nit: let's say "standard input" when talking about stdin in prose

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+				opts.Prompt = true
+			}
+
+			if !opts.IO.IsStdinTTY() {
```
I'm not sure if this is a good idea! In "scripting mode", unless there is a pseudo-tty involved, we can assume that none of stdin/stdout/stderr will be connected to a terminal (for example, a script ran from a cron job won't have a tty). However, that still doesn't mean that some input was piped to our process. If no input is piped in, then the code under this condition will block forever.

I'm not sure what would be a proper solution, though. Maybe there should be a combination of arguments/flags that signals unambiguously that there should be something expected on stdin.

vilmibm replied:
I can bring back `--json` but just as a boolean indicating whether or not to read from STDIN.

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+		}
+	}
+
+	if opts.JSON != "" {
```
Could this be `else if` to show that these branches are mutually exclusive

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+		}
+	}
+
+	if opts.Prompt {
```
Ditto `else if`

mislav commented on pkg/cmd/workflow/run/run.go:
```diff
@@ -0,0 +1,408 @@
+	path := fmt.Sprintf("repos/%s/actions/workflows/%d/dispatches",
+		ghrepo.FullName(repo), workflow.ID)
+
+	requestByte, err := json.Marshal(struct {
+		Ref    string            `json:"ref"`
+		Inputs map[string]string `json:"inputs"`
+	}{
+		Ref:    ref,
+		Inputs: providedInputs,
+	})
```
Instead of defining a new struct type inline, I think that passing a simple map would be enough!
```suggestion
	requestByte, err := json.Marshal(map[string]interface{}{
		"ref":    ref,
		"inputs": providedInputs,
	})
```

mislav commented on pkg/cmd/workflow/run/run_test.go:
```diff
@@ -0,0 +1,567 @@
+        run: |
+          echo "AUUUGH!"
+`)
+	encodedMinimalYAMLContent := base64.StdEncoding.EncodeToString(minimalYAMLContent)
```
Instead of enumerating several flavors of workflow YAMLs in this test, it might be simpler to create a more focused test just for `findInputs` and do it there

vilmibm replied:
i'm happy to add some more tests but at least in this instance the two yaml files are doing double duty...the minimal one exercises the `collectInputs` codepath when there are no inputs to collect (in addition to testing out the `on: workflow_dispatch` syntax). It's definitely sloppy and an artifact of my rushing some of this work.

mislav reviewed (dismissed):
This looks great! Thank you for all the updates. The only potential problem that I spotted is reading from non-existing stdin blocking usage in no-tty mode.

vilmibm committed:
review feedback

vilmibm committed:
review feedback

vilmibm committed:
add specific unit tests for findInputs

vilmibm committed:
correct Examples

tnturner83 commented:
main.yml
